src/services/coffee-shops.service.js

```diff
@@ -0,0 +1,50 @@
+import { getCoffeeShopsRepository } from '#repositories/coffee-shops.repository'
+import { getTokenService, refreshTokenService } from '#services/tokens.service'
+import { getSquaredEuclideanDistance } from '#utils/distance.utils'
+
+export const getCoffeeShopsService = async () => {
+  try {
+    const token = await getTokenService()
+
+    return await getCoffeeShopsRepository(token)
+  } catch (e) {
+    if (e.status === 401) {
+      // if unauthorized error, refresh the token and retry the operation
+      const newToken = await refreshTokenService()
+      return getCoffeeShopsRepository(newToken)
+    }
+
+    throw e
+  }
+}
+
+export const getClosestCoffeeShopsWithDistance = async ({
+  position,
+  limit = 3,
+}) => {
+  try {
+    const coffeeShops = await getCoffeeShopsService()
+
+    // First augment the coffee shops with the squared euclidean distance
+    const coffeeShopsWithSquaredDistance = coffeeShops.map((coffeeShop) => ({
+      ...coffeeShop,
+      squaredDistance: getSquaredEuclideanDistance(position, coffeeShop),
+    }))
```

Comment:
I like the performance optimisation here 👍. Sure, if we only had 3 shops in the list this would just result in 3 redundant operations, but in a real life situation with thousands of shops this could result in a reasonable performance boost.